File: algorithms/[1137]第 N 个泰波那契数/solution.ts
// topics = ["递归"]

/**
 * 递归 + 空间优化
 * time O(n), space O(1)
 */
function tribonacci(n: number): number {
  if (n == 0) {
    return 0;
  }
  if (n < 3) {
    return 1;
  }

  let r = 0;
  let s = 1;
  let t = 1;
  for (let i = 3; i <= n; i += 1) {
    [r, s, t] = [s, t, r + s + t];
  }
  return t;
}

File: algorithms/[1137]第 N 个泰波那契数/solution.py
# topics = ["递归"]


class Tri:
    def __init__(self):
        n = 38
        self.nums = nums = [0] * n
        nums[1] = nums[2] = 1
        for i in range(3, n):
            nums[i] = nums[i - 1] + nums[i - 2] + nums[i - 3]


class Solution:
    """
    递归 + 记忆化
    time O(1), space O(1)
    """
    t = Tri()

    def tribonacci(self, n: int) -> int:
        return self.t.nums[n]
